scapy/contrib/isotp.py

```diff
@@ -1611,3 +1613,391 @@ def recv(self, x=0xffff):
+                                                  noise_ids, True, p,
+                                                  verbose),
+                       timeout=sniff_time * 3,
+                       started_callback=send_multiple_ext(sock, extended_id,
+                                                          pkt,
+                                                          scan_block_size))
+            # without sleep it's to fast for the sock
+            time.sleep(1)
+
+        # remove duplicate IDs
+        id_list = list(set(id_list))
+        for extended_id in id_list:
+            for ext_id in range(extended_id, min(extended_id +
+                                                 scan_block_size, 256)):
+                pkt.extended_address = ext_id
+                full_id = (value << 8) + ext_id
+                sock.sniff(prn=lambda pkt: get_isotp_fc(full_id,
+                                                        return_values,
+                                                        noise_ids, True,
+                                                        pkt, verbose),
+                           timeout=sniff_time,
+                           started_callback=lambda: sock.send(pkt))
+    return return_values
+
+
+def ISOTPScan(sock, scan_range=range(0x7ff + 1), extended_addressing=False,
+              noise_listen_time=2,
+              sniff_time=0.1,
+              output_format=None,
+              can_interface="can0",
+              verbose=False):
+
+    """Scan for ISOTP Sockets on a bus and return findings
+
+    Args:
+            sock: CANSocket object to communicate with the bus under scan
+            scan_range: hexadecimal range of CAN-Identifiers to scan.
+                        Default is 0x0 - 0x7ff
+            extended_addressing: scan with ISOTP extended addressing
+            noise_listen_time: seconds to listen for default
+                               communication on the bus
+            sniff_time: time the scan waits for isotp flow control responses
+                        after sending a first frame
+            output_format: defines the format of the returned
+                           results (text, code or sockets). Provide a string
+                           e.g. "text". Default is "socket".
+            can_interface: interface used to create the returned code/sockets
+            verbose: displays information during scan
+
+    Scan for ISOTP Sockets in the defined range and returns found sockets
+    in a specified format. The format can be:
+    text: human readable output
+    code: python code for copy&paste
+    sockets: if output format is not specified, ISOTPSockets will be
+             created and returned in a list
+    """
+
+    if verbose:
+        print("Filtering background noise...")
+
+    # Send dummy packet. In most cases, this triggers activity on the bus.
+    dummy_pkt = CAN(identifier=0x123,
```

Comment:
Just for me. Why? Is this meant for SW-CAN for the Wake-up signal?

Reply:
No, ECUs usually sleep. Any communication on the CAN-Bus will cause a wake-up of the ECU.